### PR TITLE
Fix #95: shallow-clone test coverage, shallow detection, actionable warnings

### LINKED ISSUE #95: [ci] No test coverage for shallow clones, the single most likely production failure mode
## Problem

`actions/checkout` defaults to a fetch depth of 1. `findMergeBase` then returns null and Scalpel either silently builds everything under failSafe or hard-fails. There is zero coverage for this, unit or integration, although the project's own workflows (`sonarcloud.yml`, `deploy-snapshot.yml`) set `fetch-depth: 0` because they know about the trap.

This is not hypothetical: issue #45 was exactly this, a `JGitInternalException` escaping the failSafe handler on a shallow clone in Apicurio Registry CI, fixed in 0.3.9.

Related unbounded-cost problem in the same area: with `fetchBaseBranch=true` on a shallow clone, JGit 5.13's `FetchCommand` has no depth controls (shallow-fetch support arrived in JGit 6.x), so the fetch pulls the base branch's entire history over the network inside `afterProjectsRead`, on the build's critical path.

## Proposal

- Integration test cloning with depth 1 asserting the failSafe warning plus a full build.
- Unit test for `detectChanges` when no merge base exists on a shallow repository.
- Detect a shallow repository and log an actionable WARN naming `fetch-depth: 0`.
- Warn that `fetchBaseBranch` on a shallow clone fetches unbounded history on JGit 5.x.

### PR DESCRIPTION
**Pull request by paoloantinori**
## Problem

As filed in #95: shallow clones are the most likely production failure mode (`actions/checkout` defaults to fetch-depth 1; `findMergeBase` then returns null) and had zero coverage, even though the project's own workflows set `fetch-depth: 0` because they know about the trap. This is exactly what issue #45 was (JGitInternalException escaping failSafe on a shallow clone in a user's CI).

## Change

- New `ShallowCloneTest` with two fixtures: a real shallow clone via `Git.cloneRepository().setDepth(1)` against a local remote (JGit 7 has the depth API; verified it produces a real `.git/shallow`), plus a deterministic plumbing fixture (orphan branch + explicit shallow marker) kept for documentation. Both assert the merge base past the shallow boundary is unresolvable and detection terminates without throwing.
- `isShallow(Repository)`: JGit 7.7.1 exposes no public shallow accessor on Repository (only internal DepthWalk classes), so the `.git/shallow` marker check is the mechanism.
- Actionable WARN in `findMergeBase` when the merge base is null and the repo is shallow ("use fetch-depth: 0 in CI, or run git fetch --unshallow"), instead of silence.
- `fetchBranch` warns that the fetch is unbounded on shallow repos (FetchCommand.setDepth exists on JGit 7 but is not wired; noted in a comment).

## Verification

TDD (RED on the WARN, GREEN after), plus a review pass that corrected stale comments. Full `core` (133) + `extension3` (146) suites green; all fixtures offline.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of shallow Git clones when change history is incomplete.
  * Added clearer diagnostics when a merge base or historical file cannot be found.
  * Added actionable warnings explaining that additional history may be needed to restore accurate change detection.

* **Tests**
  * Added coverage for shallow clones, missing merge bases, historical file access, and full-module change detection behavior.

### CODE REVIEW

File: core/src/main/java/eu/maveniverse/maven/scalpel/core/GitChangeDetector.java [hunk 1]
```diff
@@ -49,6 +50,16 @@ public String getCurrentBranch(Repository repository) throws IOException {
...
     }
 
+    /**
+     * Tells whether the repository is shallow (created or fetched with a depth limit, e.g.
+     * actions/checkout with {@code fetch-depth: 1}), which is recorded in {@code .git/shallow}.
+     * JGit 7.x still exposes no public accessor for this ({@link Repository} has no shallow
+     * method; only the internal DepthWalk classes track it), so the marker file is checked directly.
+     */
+    public boolean isShallow(Repository repository) {
+        return repository.getDirectory() != null && new File(repository.getDirectory(), "shallow").isFile();
```
Comment (gnodet) [inline core/src/main/java/eu/maveniverse/maven/scalpel/core/GitChangeDetector.java]: **High (CI blocker):** modernizer-maven-plugin rejects `new File(repository.getDirectory(), "shallow")`. The project enforces modern Java APIs — use `Path.resolve(String)` instead:

```suggestion
    public boolean isShallow(Repository repository) {
        return repository.getDirectory() != null
                && repository.getDirectory().toPath().resolve("shallow").toFile().isFile();
    }
```

The `import java.io.File` on line 11 also becomes unnecessary.

File: core/src/test/java/eu/maveniverse/maven/scalpel/core/ShallowCloneTest.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                    .setRemote(remoteDir.toUri().toString())
+                    .setRefSpecs(new org.eclipse.jgit.transport.RefSpec("refs/heads/main:refs/heads/main"))
+                    .call();
+        }
+
+        // real depth-limited clone over a local path URI
+        Path cloneDir = tempDir.resolve("clone");
+        try (Git clone = Git.cloneRepository()
+                        .setDepth(1)
+                        .setBranchesToClone(java.util.Collections.singletonList("refs/heads/main"))
```
Comment (gnodet) [inline core/src/test/java/eu/maveniverse/maven/scalpel/core/ShallowCloneTest.java]: **High (CI blocker):** modernizer-maven-plugin rejects `Collections.singletonList(...)`. Use `List.of(...)` instead:

```suggestion
                .setBranchesToClone(List.of("refs/heads/main"))
```

**Review by gnodet — CHANGES_REQUESTED** [review]
Solid contribution addressing a real production gap (shallow clone detection and actionable warnings). The approach is well-designed: `.git/shallow` marker detection is correct given JGit 7.x exposes no public shallow accessor, and the actionable WARN messages naming both `fetch-depth: 0` and `git fetch --unshallow` directly address the user experience gap.

However, **CI is failing with 2 modernizer violations** that need fixing before merge.

**Low-severity observations (non-blocking):**

1. The `captureErr()` approach for asserting on log output is unique to this test and relies on `slf4j-simple` defaulting to `System.err`.

2. The `findMergeBase(repository, "HEAD~1", "HEAD")` assertion exercises an earlier guard (`repository.resolve` returns null) rather than the new `isShallow()` code path. The disconnected-branches fixture correctly covers the new path, so coverage is not missing — but the test comment could be clearer.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | _(unlabeled)_ | `bug` |
| Milestone | _(none)_ | `0.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by paoloantinori** [comment]
Addressed the review feedback in 0cf6610.

Modernizer violations fixed:
- GitChangeDetector.isShallow: replaced `new File(directory, "shallow").isFile()` with `Files.isRegularFile(directory.toPath().resolve("shallow"))` (Prefer java.nio.file.Path.resolve).
- ShallowCloneTest: replaced `Collections.singletonList("refs/heads/main")` with `List.of(...)` (Prefer java.util.List.of).

Non-blocking observations:
- captureErr() now carries a javadoc note documenting that it relies on slf4j-simple's default output target being System.err.
- Clarified the comment on the HEAD~1 findMergeBase assertion in the depth-1 clone test: it exercises the unresolvable-base (baseId == null) guard in findMergeBase, not isShallow.

Verified: `./mvnw -am -pl extension3 verify` (modernizer + jacoco all green) and `./mvnw -pl core test` (133 tests, 0 failures).

**Review by gnodet — APPROVED** [review]
All three previous findings are fully addressed by commit 0cf6610. Changing verdict from REQUEST_CHANGES to APPROVE.

### Previous Findings — Resolved ✅

1. **~~CI failing with 2 modernizer violations~~** — Fixed. Replaced `new File(...).isFile()` with `Files.isRegularFile(Path)` and `Collections.singletonList()` with `List.of()`. CI now passes on all platforms and JDK versions.

2. **~~captureErr() relies on slf4j-simple~~** — Addressed. Added a Javadoc to `captureErr()` explicitly documenting the assumption.

3. **~~findMergeBase test comment unclear~~** — Addressed. Expanded the inline comment to clarify that it exercises the unresolvable-base guard (`baseId == null` early return), not `isShallow` itself.

### General Comments

- Clean, focused fix commit (+12/-4 lines) addressing all feedback in a single commit.
- CI fully green across all 9 matrix entries.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | _(unlabeled)_ | `bug` |
| Milestone | _(none)_ | `0.4.0` |

---

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: core/src/main/java/eu/maveniverse/maven/scalpel/core/GitChangeDetector.java (hunk 1 above)
Comment (paoloantinori) [reply core/src/main/java/eu/maveniverse/maven/scalpel/core/GitChangeDetector.java]: Fixed in 0cf6610: isShallow now uses Files.isRegularFile(repository.getDirectory().toPath().resolve("shallow")); verify runs modernizer clean.

File: core/src/test/java/eu/maveniverse/maven/scalpel/core/ShallowCloneTest.java (hunk 2 above)
Comment (paoloantinori) [reply core/src/test/java/eu/maveniverse/maven/scalpel/core/ShallowCloneTest.java]: Fixed in 0cf6610: Collections.singletonList replaced with List.of; the modernizer violations are gone from the verify build.

**Comment by paoloantinori** [comment]
Addressed in 02dfdba.

`findMergeBase` now checks `isShallow` on the unresolved-revision early returns as well: when the base or head revision fails to resolve in a shallow repository, the fetch-depth: 0 / `git fetch --unshallow` remediation is logged before the generic unresolved-revision warning, instead of being skipped. The guidance text lives in a single `warnIfShallow` helper shared with the existing no-merge-base check, so the both-revisions-resolve path is unchanged and the message is never emitted twice on one call.

`ShallowCloneTest.realShallowClone_isShallowAndCannotResolvePreCloneHistory` (the real `setDepth(1)` clone fixture, where `HEAD~1` is beyond the boundary) now captures stderr around `findMergeBase(HEAD~1, HEAD)` and asserts the shallow cause and the fix are present in the output. Both suites green: core 133/133, extension3 146/146.

**Review by gnodet — APPROVED** [review]
## Re-review (1 new commit: shallow guidance on unresolved revisions)

Clean refactoring that extracts the shallow-clone warning into a reusable `warnIfShallow` method and extends it to cover unresolved-revision early returns, with matching test coverage that asserts the warning content.

### Observations (non-blocking)

- Minor inconsistency in warning order: for `baseId == null` / `headId == null`, `warnIfShallow` fires before the specific "Cannot resolve" message (explanation then symptom); for `mergeBase == null`, the specific message fires first (symptom then explanation). Not blocking — both orderings work fine.
- The `warnIfShallow` extraction eliminates duplicated warning text and is reusable across all three failure paths — good refactoring.
- Test captures stderr and asserts both cause ("Repository is shallow") and remediation ("fetch-depth: 0" / "--unshallow"), providing solid behavioral coverage.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Milestone | _(none)_ | `0.4.0` |

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — APPROVED** [review]
Well-constructed PR that fills a real production gap (zero coverage for shallow clones) with proper detection, actionable warnings, and thorough tests. CI fully green across all 9 matrix entries.

**Observations (non-blocking):**
- After merge with main (which added `MissingObjectException` handling in #144), the `MissingObjectException` catch block won't call `warnIfShallow()`. The common production case (`actions/checkout` depth=1 where `.git/shallow` exists) is covered via the `mergeBase==null` path, but a follow-up adding `warnIfShallow` to the `MissingObjectException` catch would provide actionable guidance for the edge case.
- The `captureErr()` helper redirects `System.err` and is not thread-safe — works because Surefire runs tests sequentially. A future migration to a log-capturing test extension (e.g., LogCaptor) would be more robust.
- Good test design: real `setDepth(1)` clone test validates the full JGit shallow-clone stack, while the disconnected-roots fixture provides a deterministic alternative.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Milestone | _(none)_ | `0.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by gnodet** [comment]
The `MissingObjectException` catch block should also call `warnIfShallow()` — please add it in this PR rather than leaving it as a follow-up. The shallow-clone guidance is the whole point of this change, and the `MissingObjectException` path is a real scenario on depth-limited clones. No reason to merge with a known gap and track it separately.

**Comment by paoloantinori** [comment]
gnodet's request addressed in commit 0f28df7:

The MissingObjectException catch blocks now call warnIfShallow() before propagating the generic warning and (for findMergeBase) returning null. Both catch paths emit the actionable shallow-clone remediation:

- `findMergeBase`: catch MissingObjectException → warnIfShallow → log generic "commit history is incomplete" → return null (aligns with upstream main's behavior, preserving failSafe semantics).

- `readFileAtCommit`: catch MissingObjectException → warnIfShallow → rethrow (semantics preserved: missing commit is different from file-not-found).

New test `realShallowClone_missingObjectBeyondBoundaryEmitsShallowGuidance` exercises both code paths on the real depth-1 clone fixture, asserting that the shallow cause ("Repository is shallow") and the fix ("fetch-depth: 0" / "--unshallow") are present in the logged output.

Both suites green: core 134/134, extension3 146/146.

**Review by gnodet — APPROVED** [review]
Re-review after latest commit (0f28df7): correctly addresses the review request to add `warnIfShallow()` to `MissingObjectException` catch blocks in both `findMergeBase` and `readFileAtCommit`, with comprehensive test coverage.

Observations:

1. **Merge conflict with main** (medium) — main's `findMergeBase` (from PR #144) already has `MissingObjectException` + `IOException` catch blocks without `warnIfShallow`. This PR's version has `warnIfShallow` but lacks the `IOException` catch block. Resolution is straightforward: combine both. Not a code quality issue, but blocks merge.

2. **warnIfShallow parameter semantics** (low) — in `readFileAtCommit`, `warnIfShallow(repository, commitId.getName(), path)` passes a file path as the second parameter, but `warnIfShallow`'s message template references `{baseBranch}` and `{head}`. The resulting message is slightly semantically off (path is not a branch ref), though still actionable for users.

The new test (`realShallowClone_missingObjectBeyondBoundaryEmitsShallowGuidance`) exercises both code paths with a genuine depth-1 clone — well-designed test that validates behavioral correctness.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
